supervisor/docker/network.py

```diff
@@ -42,55 +68,65 @@ def containers(self) -> list[str]:
     @property
     def gateway(self) -> IPv4Address:
         """Return gateway of the network."""
-        return DOCKER_NETWORK_MASK[1]
+        return DOCKER_IPV4_NETWORK_MASK[1]
 
     @property
     def supervisor(self) -> IPv4Address:
         """Return supervisor of the network."""
-        return DOCKER_NETWORK_MASK[2]
+        return DOCKER_IPV4_NETWORK_MASK[2]
 
     @property
     def dns(self) -> IPv4Address:
         """Return dns of the network."""
-        return DOCKER_NETWORK_MASK[3]
+        return DOCKER_IPV4_NETWORK_MASK[3]
 
     @property
     def audio(self) -> IPv4Address:
         """Return audio of the network."""
-        return DOCKER_NETWORK_MASK[4]
+        return DOCKER_IPV4_NETWORK_MASK[4]
 
     @property
     def cli(self) -> IPv4Address:
         """Return cli of the network."""
-        return DOCKER_NETWORK_MASK[5]
+        return DOCKER_IPV4_NETWORK_MASK[5]
 
     @property
     def observer(self) -> IPv4Address:
         """Return observer of the network."""
-        return DOCKER_NETWORK_MASK[6]
+        return DOCKER_IPV4_NETWORK_MASK[6]
+
+    def _get(self, network_id: str) -> docker.models.networks.Network:
+        """Get a network by ID."""
+        return self.docker.networks.get(network_id)
 
-    def _get_network(self) -> docker.models.networks.Network:
+    def _create(self, **kwargs: Any) -> docker.models.networks.Network:
+        """Create a new network."""
+        return self.docker.networks.create(**kwargs)
+
+    def _get_network(self, enable_ipv6: bool = False) -> docker.models.networks.Network:
         """Get supervisor network."""
         try:
-            return self.docker.networks.get(DOCKER_NETWORK)
+            if network := self._get(DOCKER_NETWORK):
+                if network.attrs.get(DOCKER_ENABLEIPV6) == enable_ipv6:
+                    return network
+                try:
+                    network.remove()
+                    _LOGGER.info(
+                        "Migrating Supervisor network to %s",
+                        "IPv4/IPv6 Dual-Stack" if enable_ipv6 else "IPv4-Only",
+                    )
+                except docker.errors.APIError:
+                    _LOGGER.info(
+                        "Supervisor restart is required to apply network changes"
+                    )
```

Comment:
So a Supervisor restart is not enough, since we leave containers running in that case. At the very least it needs a system restart, so:

```suggestion
                    _LOGGER.info(
                        "Migrating Supervisor network to %s",
                        "IPv4/IPv6 Dual-Stack" if enable_ipv6 else "IPv4-Only",
                    )
                    network.remove()
                except docker.errors.APIError:
                    _LOGGER.warning(
                        "System restart is required to apply Supervisor network changes"
                    )
```

But I think even that is not enough: The Observer runs with restart policy always, which makes it run on startup, probably even before Supervisor :cry: 

I think we need to really add some special code here: Check if Observer plug-in is running, if so either remove and reattach to the hassio network, or simply stop it and rely on later code to start it again.

Reply:
So we need to disconnect containers from the network. Something like:
```
                except docker.errors.APIError:
                    if (containers := self.docker.networks.containers):
                        for container in containers:
                            self.docker.networks.disconnect(container)
```
?

Reply:
Something like this?